### PR TITLE
Create ProviderStore interface, migrate controlplane to use it

### PR DESCRIPTION
**Pull request by dmjb**
Relates to #2845

This provides an interface for retrieving Providers out of the database, and encapsulates the logic for doing so. This makes it easier to stub out the provider retrieval logic for testing, and decouples the controlplane from the details of finding the right provider for a request.

There are other places in the codebase where this should be used, but that is an exercise for a future PR.

# Summary

***Provide a brief overview of the changes and the issue being addressed. 
Explain the rationale and any background necessary for understanding the changes. 
List dependencies required by this change, if any.***

Fixes #(related issue)

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [x] Checked that related changes are merged.

### CODE REVIEW

File: internal/providers/store.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright 2024 Stacklok, Inc
```
Comment (dmjb) [inline internal/providers/store.go:1]: Similar to my other PR - this relies on the existing tests in the controlplane to validate it. It should get its own dedicated test suite in due course.

File: internal/providers/store.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/stacklok/minder/internal/db"
+	"github.com/stacklok/minder/internal/util"
+)
+
+// ProviderStore provides methods for retrieving Providers from the database
+type ProviderStore interface {
```
Comment (dmjb) [inline internal/providers/store.go:32]: Often the term `Repository` is used for these types of interface. However, given that that word already has an important meaning in the minder codebase, I am reusing `Store`

**Review by evankanderson — DISMISSED** [review]
A bunch of small nits, but this seems like a straightforward step-along-the-way refactor.

File: internal/providers/store.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	// GetByName returns the provider instance in the database as identified
+	// by its project ID and name.
+	GetByName(ctx context.Context, projectID uuid.UUID, name string) (*db.Provider, error)
```
[the remark this answers is not in the document]
Comment (dmjb) [reply internal/providers/store.go:35]: Yes.

**Comment by dmjb** [comment]
Replaced by this PR: https://github.com/stacklok/minder/pull/2900

**Pull request closed without merging**